### ISSUE TITLE
support CSS color level 5 functions

### ISSUE THREAD

**Issue by stevenvachon** [sass/sass#2078]
``` css
element {
  border: 1px solid color(black alpha(50%));
}
```

It currently throws:

```
$color: 50% is not a color for `alpha'
```

Now, I don't expect Sass to do anything with this, as that's the job of a post-processor.. but Sass should at least not throw an error.

**Comment by andypinet** [comment]
+1

**Comment by douglasdeodato** [comment]
Have you tried use sass palettes? its a nice feature and I think can help you on that.

you can use: saturate, desaturate, lighten, darken to match your colours, as well you can usex rgba,rgb, hex.

see that:  https://gist.github.com/douglasdeodato/09e64152c8f3a81f0bea794f88e29c39

not sure if that resolve your problem.

**Comment by stevenvachon** [comment]
Yes, that is what I am transitioning away from. I want to write mostly standard CSS.

**Comment by douglasdeodato** [comment]
ok, sorry maybe that should work?
- (Sass::Script::Value::Color) rgba($color, $alpha)
  Sets the opacity of an existing color.

Examples:

rgba(#102030, 0.5) => rgba(16, 32, 48, 0.5)
rgba(blue, 0.2)    => rgba(0, 0, 255, 0.2)

https://gist.github.com/douglasdeodato/6b7951c2090e673a526324ceaac4c56f

**Comment by stevenvachon** [comment]
http://cssnext.io

**Comment by xzyfer** [comment]
If I understand correctly this is referring to the CSS Color Module Level 4 (not 3) [`color` function](https://drafts.csswg.org/css-color-4/#color-function)?

This spec appears to still be an editors draft which means it has the potential to change. Sass traditionally does not implement new specifications before they're stable (read: implemented in a browser) because if the spec changes we're stuck supporting the old and stable versions for a long time.

Technically it is possible for Sass to opt out of parsing `color()` function arguments like it does for `calc()` once we're sure the spec isn't going to change the function name under our feet. Maybe @tabatkins can speak to stability?

**Comment by tabatkins** [comment]
And in particular, `color()` did change - what was previously there now lives at `adjust-color()` (and has an issue noting that it needs some serious changes), and the current `color()` is about referring to a color in an arbitrary colorspace defined by a `@color-profile` rule.

I highly recommend not trying to copy whatever we're doing in this case until browsers start shipping it.  (That's a pretty good rule in general; it's not the spec status that matters, it's the shipping implementations.)

**Comment by stevenvachon** [comment]
Please re-read my original post:

> Now, I don't expect Sass to do anything with this, as that's the job of a post-processor.. but **Sass should at least not throw an error.**

**Comment by stevenvachon** [comment]
A possible solution is providing the option to prefix all Sass functions:

``` css
color: sass-darken(black, 50%);
```

**Comment by chriseppstein** [comment]
> A possible solution is providing the option to prefix all Sass functions

This is a feature of Sass 4.0.

In Sass 3.5, there will be a generic work around that could be done in the stylesheet, without needing a patch release.

In sass 3.4 (current release), we will need to do a patch release to handle this.

@tabatkins There's a ton of functions here that will stomp all over a bunch of Sass functions and probably some user-defined ones. That's fine, obv, but what is the current browser support for this new spec? I'm surprised users are hitting this already.

**Comment by chriseppstein** [comment]
@stevenvachon According to @tabatkins this spec is still very much in flux, especially around the `color()` function that you're using here. What is your current use case?

**Comment by stevenvachon** [comment]
@chriseppstein I'm currently using Less followed by  [cssnext](https://www.npmjs.com/package/postcss-cssnext). I'd like to swap out Less for Sass/SCSS, but currently cannot because I get errors.

**Edit:**
Without [less-plugin-future-compat](https://www.npmjs.com/less-plugin-future-compat), Less will error as well. I had tried to do the same for libsass ([libsass#1911](https://github.com/sass/libsass/issues/1911)), but it wasn't possible.

**Comment by stevenvachon** [comment]
@chriseppstein ?

It would be awesome if 3.4.x could get this feature as an option.

**Comment by chriseppstein** [comment]
@stevenvachon In Sass 3.5 there is a good work-around that is already landed. In Sass 3.4 you can do:

``` scss
@function css-alpha($args...) {
  @return unquote("alpha(#{$args})");
}
element {
  border: 1px solid color(black css-alpha(50%));
}
```

I know it's not ideal. But it does produce the desired output.

**Comment by stevenvachon** [comment]
Since I already have the desired code using Less, I'll stick with that for now. When is 3.5 due out?

**Comment by chriseppstein** [comment]
@stevenvachon we'll announce on twitter when it's released.

**Comment by stevenvachon** [comment]
Any time soon, perhaps?

With the above information, should we close this then?

**Comment by chriseppstein** [comment]
@stevenvachon not in the next month. I'd like to leave this open for now.

**Comment by stevenvachon** [comment]
@chriseppstein is there an issue or something for the function prefixing in v4 that we can follow?

**Comment by imkremen** [comment]
@chriseppstein , @xi is an author of library that implement some part of color functions for SASS, I think that for now (after some redesign) it can be useful as polyfill for color level 4 functions.

**Comment by brookjordan** [comment]
I would love for CSS Colors Level 4 to be supported.
I have so many printed strings all over the place now just to get this working.

Using `rgb(0 0 0 / 0.5)` or `rgb(0, 0, 0, 0.5)` shouldn't throw an error.

**Comment by nex3** [comment]
@brookjordan Our policy is to wait until at least one browser supports a new CSS feature to begin supporting it in Sass. I believe (according to https://github.com/Fyrd/caniuse/issues/2864) that no browsers support this yet.

**Comment by brookjordan** [comment]
Working in Firefox and Chrome:
```css
body {
  color: rgb(0 0 0 / 0.5);
/* and */
  background-color: rgb(0, 0, 0, 0.5);
}
```

**Comment by nex3** [comment]
Great! Since only that part of Colors Level 4 has been implemented in browsers, I've filed https://github.com/sass/sass/issues/2564 to track it separately.

**Comment by mirisuzanne** [comment]
Color adjustment functions have been moved to [Level 5](https://drafts.csswg.org/css-color-5/)
